### PR TITLE
Chemtank Overhaul

### PR DESCRIPTION
**Pull request by Spacedong**
:cl: blute
add: Chemical Tank Overhaul!
imageadd: Chemical tanks now accurately display the liquid inside them.
add: Chemical tanks filled with fuel now explode more violently when ruptured.
add: Fire extinguishers can now accept any chemical.
del: Removed High-Capacity watertanks.
/:cl:

Ports the chemtank part of https://github.com/HippieStationCode/HippieStation13/pull/1181

### CODE REVIEW

**Comment by Spacedong** [comment]
>CONFLICTS BOY IF YOU DONT

**Comment by tortellinitony** [comment]
No thanks.

File: code/__DEFINES/flags.dm [hunk 1]
```diff
@@ -23,6 +23,10 @@
 
 #define NOSLIP		1024 		//prevents from slipping on wet floors, in space etc (NOTE: flag shared with THICKMATERIAL for external suits and helmet)
 
+#define OPENCONTAINER		4096 //is an open container for chemistry purposes
+
+#define INJECTONLY		8192
```
Comment (tortellinitony) [inline code/__DEFINES/flags.dm:28]: No, do not make this a flag. Make it a secondary flag or a variable that works with opencontainer please.

File: code/modules/hippie/code/modules/reagent/reagent_dispenser.dm [hunk 2]
```diff
@@ -34,4 +34,3 @@
 	desc = "A machine filled with the clown's thick juice! NICE!"
 	icon = 'icons/hippie/obj/vending.dmi'
 	icon_state = "honk_cooler"
-	reagent_id = "banana"
```
Comment (tortellinitony) [inline code/modules/hippie/code/modules/reagent/reagent_dispenser.dm:37]: nah mate fuck off

File: code/modules/reagents/reagent_dispenser.dm [hunk 3]
```diff
@@ -1,196 +1,340 @@
...
+	icon_state = "honk_cooler"
 	anchored = 1
-	reagent_id = "virusfood"
\ No newline at end of file
+	var/cups = 50
+	brokenvars = list("honk-cooler", "honk_cooler", /obj/structure/reagent_dispensers/honk_cooler)
+
+/obj/structure/reagent_dispensers/honk_cooler/New(loc, empty = 0)
+	..()
+	if(!empty) reagents.add_reagent("banana",500)
```
Comment (tortellinitony) [inline code/modules/reagents/reagent_dispenser.dm:300]: no why the hell

File: code/__DEFINES/flags.dm [hunk 4]
```diff
@@ -23,6 +23,10 @@
 
 #define NOSLIP		1024 		//prevents from slipping on wet floors, in space etc (NOTE: flag shared with THICKMATERIAL for external suits and helmet)
 
+#define OPENCONTAINER		4096 //is an open container for chemistry purposes
```
Comment (ghost) [inline code/__DEFINES/flags.dm:26]: This flag was removed, there is nothing that checks for it, this won't even work

File: code/__DEFINES/flags.dm (hunk 1 above)
Comment (ghost) [inline code/__DEFINES/flags.dm:28]: This flag was removed, there is nothing that checks for it, this won't even work

File: code/modules/reagents/reagent_dispenser.dm [hunk 5]
```diff
@@ -1,196 +1,340 @@
...
+	icon = 'icons/obj/chemtank.dmi'
+	icon_state = "chemtank"
 	density = 1
 	anchored = 0
 	pressure_resistance = 2*ONE_ATMOSPHERE
-	obj_integrity = 300
-	max_integrity = 300
-	var/tank_volume = 1000 //In units, how much the dispenser can hold
-	var/reagent_id = "water" //The ID of the reagent that the dispenser uses
+	var/list/brokenvars = list("storage tank", "chemtank", /obj/structure/reagent_dispensers)// when it blows up,what will it become,made it qdel src and make a statuelike thing to avoid having to change syringes,droppers,glasses etc with snowflakey checks
```
Comment (ghost) [inline code/modules/reagents/reagent_dispenser.dm:9]: There is no need to use a `list` for this, use `initial()`

File: code/modules/reagents/reagent_dispenser.dm [hunk 6]
```diff
@@ -1,196 +1,340 @@
...
+	var/exploded = 0 // if atleast one reaction happened in boom(), this becomes 1
+	var/list/bypassUIicons = list( // list which have all the icons where reagent overlay and update_icon proc in general must not work
```
Comment (ghost) [inline code/modules/reagents/reagent_dispenser.dm:11]: Please do not initialize lists directly on the object, or in `/new` please use the `LAZY_` list helpers when working with lists.

More about that can be found [here](https://github.com/HippieStation/HippieStation13/blob/master/code/__HELPERS/lists.dm#L459)

**ALSO**

`bypassUIicons` is a bad variable name, change this into `lower_case_` also just override `update_icon` and check it in there for these specific objects

File: code/modules/reagents/reagent_dispenser.dm [hunk 7]
```diff
@@ -1,196 +1,340 @@
...
-	create_reagents(tank_volume)
-	reagents.add_reagent(reagent_id, tank_volume)
+/obj/structure/reagent_dispensers/attack_hand()
 	..()
+	update_icon()
 
-/obj/structure/reagent_dispensers/examine(mob/user)
-	..()
+/obj/structure/reagent_dispensers/update_icon()
+	overlays.Cut()
```
Comment (ghost) [inline code/modules/reagents/reagent_dispenser.dm:27]: **_Please do not manually change the overlays_**

Use `/proc/cut_overlays()` and `/proc/add_overlay(image, priority = 0)` so that overlay changes do not cause any lag

File: code/modules/reagents/reagent_dispenser.dm [hunk 8]
```diff
@@ -1,196 +1,340 @@
...
+				icon_state = "weldtank"
+			if("water")
+				name = "watertank"
+				desc = "It can hold large amount of chemicals. Mainly used to hold water."
+				icon_state = "watertank"
+			else
+				name = "storage tank"
+				desc = "It can hold large amount of chemicals."
+				icon_state = "chemtank"
+		var/image/filling = image('icons/obj/chemtank.dmi', src, "tankfilling1")
```
Comment (ghost) [inline code/modules/reagents/reagent_dispenser.dm:43]: Please cache this image

File: code/modules/reagents/reagent_dispenser.dm [hunk 9]
```diff
@@ -1,196 +1,340 @@
...
+		var/percent = round((reagents.total_volume / reagents.maximum_volume) * 100)
+		switch(percent)
+			if(0 to 25)			filling.icon_state = "tankfilling1"
+			if(25 to 50)		filling.icon_state = "tankfilling2"
+			if(50 to 75)		filling.icon_state = "tankfilling3"
+			if(75 to INFINITY)	filling.icon_state = "tankfilling4"
+		filling.color = mix_color_from_reagents(reagents.reagent_list)
+		overlays += filling
```
Comment (ghost) [inline code/modules/reagents/reagent_dispenser.dm:51]: **_Please do not manually change the overlays_**

Use `/proc/cut_overlays()` and `/proc/add_overlay(image, priority = 0)` so that overlay changes do not cause any lag

File: code/modules/reagents/reagent_dispenser.dm [hunk 10]
```diff
@@ -1,196 +1,340 @@
...
+	if(!(flags & INJECTONLY))//here we check the lid!we want it to appear if it's closed.
+		var/image/lid = image('icons/obj/chemtank.dmi', src, "tanklid")
```
Comment (ghost) [inline code/modules/reagents/reagent_dispenser.dm:53]: Please cache this image

File: code/modules/reagents/reagent_dispenser.dm [hunk 11]
```diff
@@ -1,196 +1,340 @@
...
+		overlays += lid
```
Comment (ghost) [inline code/modules/reagents/reagent_dispenser.dm:54]: **_Please do not manually change the overlays_**

Use `/proc/cut_overlays()` and `/proc/add_overlay(image, priority = 0)` so that overlay changes do not cause any lag

File: code/modules/reagents/reagent_dispenser.dm [hunk 12]
```diff
@@ -1,196 +1,340 @@
...
+		user.attack_log += text("\[[time_stamp()]\] <font color='red'>Has detonated a chem tank @ [src ? "[x],[y],[z]" : "UNKNOWN LOCATION"]</font>")
+	if(reagents && reagents.reagent_list.len && !exploded)
+		reagents.chem_temp = 1000
+		reagents.handle_reactions()
+		if(!src) return 1// safety check,explosion could've deleted it!
+		exploded = 1
+		var/volumepereffect = reagents.total_volume / 9
+		var/list/effectlist = list()
+		visible_message("<span class='danger'>[src] ruptures, spraying its contents everywhere!</span>")
+		spawn(-1)
```
Comment (ghost) [inline code/modules/reagents/reagent_dispenser.dm:70]: Please use the `addtimer` and `callback` system to do this instead of using spawns, you can find more information about this [here](https://github.com/HippieStation/HippieStation13/blob/master/code/datums/callback.dm)

File: code/modules/reagents/reagent_dispenser.dm [hunk 13]
```diff
@@ -1,196 +1,340 @@
...
+			for(var/i in 1 to 11) // i will be later used for direction. i'm a smart man am i not?admit it fucker
```
Comment (ghost) [inline code/modules/reagents/reagent_dispenser.dm:71]: I'm quite sure there is a proc that gets all possible directions and/or execute something for all directions

File: code/modules/reagents/reagent_dispenser.dm [hunk 14]
```diff
@@ -1,196 +1,340 @@
...
+				if(i == 3 || i == 7) continue // those aren't directions.
+				var/obj/effect/particle_effect/water/W = new /obj/effect/particle_effect/water(get_turf(src))
+				W.create_reagents(200) // maximum it should hold is 200
+				reagents.trans_to(W,volumepereffect)
+				W.color = mix_color_from_reagents(W.reagents.reagent_list)
+				if(i != 11) // 11 is for the effect on the same tile of the src
+					step(W,i)
+				for(var/atom/B in get_turf(W))
```
Comment (ghost) [inline code/modules/reagents/reagent_dispenser.dm:79]: Please use _istypeless for loops_ you can find more information about this in the [contributing.md](https://github.com/HippieStation/HippieStation13/blob/master/CONTRIBUTING.md)

File: code/modules/reagents/reagent_dispenser.dm [hunk 15]
```diff
@@ -1,196 +1,340 @@
...
+					if(B == W || B.invisibility) //we ignore the effect itself and stuff below the floor
+						continue
+					W.reagents.reaction(B, VAPOR)
+				W.reagents.reaction(get_turf(W), TOUCH)
+				effectlist += W
+			for(var/i in 1 to effectlist.len)
+				sleep(10)
```
Comment (ghost) [inline code/modules/reagents/reagent_dispenser.dm:86]: Please use the `addtimer` and `callback` system to do this instead of using spawns, you can find more information about this [here](https://github.com/HippieStation/HippieStation13/blob/master/code/datums/callback.dm)

File: code/modules/reagents/chemistry/reagents/other_reagents.dm [hunk 16]
```diff
@@ -859,6 +859,25 @@
 	M.adjustToxLoss(1, 0)
 	. = 1
 	..()
+/datum/reagent/fuel/on_mob_life(mob/living/M)
```
Comment (tortellinitony) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm:862]: Why is this here?

File: code/modules/reagents/chemistry/reagents/other_reagents.dm [hunk 17]
```diff
@@ -859,6 +859,25 @@
...
+	M.adjustToxLoss(1)
+	..()
+
+/datum/chemical_reaction/fuel_explosion
```
Comment (tortellinitony) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm:866]: Please follow the format of other explosion interactions

File: code/modules/reagents/reagent_dispenser.dm [hunk 18]
```diff
@@ -1,196 +1,340 @@
...
-	icon = 'icons/obj/objects.dmi'
-	icon_state = "water"
+	name = "storage tank"
+	desc = "It can hold large amount of chemicals."
+	icon = 'icons/obj/chemtank.dmi'
+	icon_state = "chemtank"
 	density = 1
 	anchored = 0
 	pressure_resistance = 2*ONE_ATMOSPHERE
-	obj_integrity = 300
```
Comment (tortellinitony) [inline code/modules/reagents/reagent_dispenser.dm:9]: Why are you getting rid of these?

File: code/modules/hippie/code/modules/reagent/reagent_dispenser.dm [hunk 19]
```diff
@@ -3,7 +3,7 @@
 		if(!user.drop_item())
 			return
 		qdel(I)
-		paper_cups++
+		cups++
```
Comment (Pyko1) [inline code/modules/hippie/code/modules/reagent/reagent_dispenser.dm]: Too bad the var is still called paper_cups, and now you are just trying to something completely different that isn't even defined.

**Comment by Pyko1** [comment]
This is not even a port, its a downgrade to old code so we can have few old features.

Port them properly to the current code, or it will be pain in the ass to keep up if tg changes crap as well.

**Comment by banthebantz** [comment]
good feature, bad port. dong pls
